orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -0,0 +1,412 @@
+        break;
+      case STRING:
+        orcType = TypeDescription.createString();
+        break;
+      case UUID:
+      case FIXED:
+      case BINARY:
+        orcType = TypeDescription.createBinary();
+        break;
+      case DECIMAL: {
+        Types.DecimalType decimal = (Types.DecimalType) type;
+        orcType = TypeDescription.createDecimal()
+            .withScale(decimal.scale())
+            .withPrecision(decimal.precision());
+        break;
+      }
+      case STRUCT: {
+        orcType = TypeDescription.createStruct();
+        for (Types.NestedField field : type.asStructType().fields()) {
+          TypeDescription childType = toOrc(field.fieldId(), field.type(), field.isRequired());
+          orcType.addField(field.name(), childType);
+        }
+        break;
+      }
+      case LIST: {
+        Types.ListType list = (Types.ListType) type;
+        TypeDescription elementType = toOrc(list.elementId(), list.elementType(),
+            list.isElementRequired());
+        orcType = TypeDescription.createList(elementType);
+        break;
+      }
+      case MAP: {
+        Types.MapType map = (Types.MapType) type;
+        TypeDescription keyType = toOrc(map.keyId(), map.keyType(), true);
+        TypeDescription valueType = toOrc(map.valueId(), map.valueType(), map.isValueRequired());
+        orcType = TypeDescription.createMap(keyType, valueType);
+        break;
+      }
+      default:
+        throw new IllegalArgumentException("Unhandled type " + type.typeId());
+    }
+
+    // Set Iceberg column attributes for mapping
+    orcType.setAttribute(ICEBERG_ID_ATTRIBUTE, String.valueOf(fieldId));
+    orcType.setAttribute(ICEBERG_REQUIRED_ATTRIBUTE, String.valueOf(isRequired));
+    return orcType;
+  }
+
+  /**
+   * Converts an Iceberg schema to a corresponding ORC schema within the context of an existing
+   * ORC file schema. This method also handles schema evolution from the original ORC file schema
+   * to the given Iceberg schema.
+   *
+   * @param schema an Iceberg schema
+   * @param originalOrcSchema an existing ORC file schema
+   * @return the resulting ORC schema
+   */
+  static TypeDescription toOrc(Schema schema, TypeDescription originalOrcSchema) {
+    final Map<Integer, OrcColumn> icebergToOrc = icebergToOrcMapping("root", originalOrcSchema);
+    return toOrc(Integer.MIN_VALUE, schema.asStruct(), true, icebergToOrc);
+  }
+
+  private static TypeDescription toOrc(Integer fieldId, Type type, boolean isRequired,
+                                       Map<Integer, OrcColumn> mapping) {
+    final TypeDescription orcType;
+
+    switch (type.typeId()) {
+      case STRUCT: {
+        orcType = TypeDescription.createStruct();
+        for (Types.NestedField nestedField : type.asStructType().fields()) {
+          String name = Optional.ofNullable(mapping.get(nestedField.fieldId()))
+              .map(OrcColumn::name)
+              .orElse(nestedField.name());
```

Comment:
if the mapping is missing, it implies that the file schema does not have that field, but the projected schema does, in that case we'd need to provide a default value right? similar to Avro? . Does ORC take care of that automatically, maybe we should have a test for this if not already .

@rdblue, maybe org.apache.iceberg.avro.TestReadProjection can we made taken out of the `org.apache.iceberg.avro` package to be extended and used for all formats? 

I'm not suggesting making these changes now, just adding notes for future tracking

Reply:
`TestReadProjection` in iceberg-data is a generic version of this that you can implement. I agree that those tests should be added before we mark this no longer experimental.

Reply:
I agree. I'll work on this.

Reply:
Update: I have implemented these tests with ORC and I have 3 failing, so I'm working on them to be fully compliant with projection tests. I should be able to fix them during this week. This would be the only remaining item.

Reply:
Sounds great! Thanks for the update!